dhis-2/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/security/SessionController.java

```diff
@@ -0,0 +1,119 @@
+import org.hisp.dhis.common.HashUtils;
+import org.hisp.dhis.common.OpenApi;
+import org.hisp.dhis.feedback.NotFoundException;
+import org.hisp.dhis.security.RequiresAuthority;
+import org.hisp.dhis.user.User;
+import org.hisp.dhis.user.UserDetails;
+import org.hisp.dhis.user.UserService;
+import org.hisp.dhis.webapi.mvc.annotation.ApiVersion;
+import org.springframework.security.core.session.SessionInformation;
+import org.springframework.security.core.session.SessionRegistry;
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.PathVariable;
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RestController;
+
+/**
+ * @author Morten Svanæs <msvanaes@dhis2.org>
+ */
+@OpenApi.Document(
+    entity = User.class,
+    classifiers = {"team:platform", "purpose:support"})
+@RestController
+@RequestMapping("/api/sessions")
+@RequiredArgsConstructor
+@ApiVersion({DhisApiVersion.DEFAULT, DhisApiVersion.ALL})
+public class SessionController {
+
+  private final SessionRegistry sessionRegistry;
+  private final UserService userService;
+
+  @GetMapping(produces = APPLICATION_JSON_VALUE)
+  @RequiresAuthority(anyOf = ALL)
+  public Map<String, String> listAllSessions() throws NotFoundException {
+    return listAllUserSessions().stream()
+        .collect(
+            Collectors.toMap(
+                s -> HashUtils.hashSHA1(s.getSessionId().getBytes()),
+                s ->
+                    ((UserDetails) s.getPrincipal()).getUsername()
+                        + (s.isExpired() ? ", (inactive)" : ", (active)")));
+  }
+
+  private List<SessionInformation> listAllUserSessions() throws NotFoundException {
+    List<SessionInformation> allSessions = new ArrayList<>();
+    List<User> allUsers = userService.getAllUsers();
+    for (User user : allUsers) {
+      allSessions.addAll(
+          sessionRegistry.getAllSessions(userService.createUserDetails(user.getUid()), true));
+    }
+    return allSessions;
+  }
+
+  @PostMapping(value = "/{username}/invalidate", produces = APPLICATION_JSON_VALUE)
+  @RequiresAuthority(anyOf = ALL)
+  public void invalidateSessions(@PathVariable("username") String username) {
+    User user = userService.getUserByUsername(username);
+    UserDetails userDetails = userService.createUserDetails(user);
+    if (userDetails != null) {
+      List<SessionInformation> allSessions = sessionRegistry.getAllSessions(userDetails, false);
+      allSessions.forEach(SessionInformation::expireNow);
+    }
+  }
+
+  @PostMapping(value = "/invalidateAll", produces = APPLICATION_JSON_VALUE)
+  @RequiresAuthority(anyOf = ALL)
+  public void invalidateAllSessions() {
+    invalidateAllUserSessions();
+  }
+
+  private void invalidateAllUserSessions() {
+    List<User> allUsers = userService.getAllUsers();
+    for (User user : allUsers) {
```

Comment:
This seems like a unnecessary cumbersome way to invalidate all sessions. The service called from the controller should have this as a dedicated operation as it likely can just do something similar to a  `Map.clear()` to perform it instead of fetching and looping stuff here.

Reply:
I know, but it's because the internals of the Redis based session store lack a way of looping through all sessions.

Reply:
just for reference, the most users in a system I've heard of is 10k. Depending how often this method is called, it could impact the system.

Reply:
I don't think this is going to be an issue, the only use-case right now is the LoginTest to invalidate sessions after each test. It's also only possible to use by ALL authority.